metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala

```diff
@@ -1,65 +1,137 @@
+    mutable.Map[FullyQualifiedClassName, TestEntry]()
+
+  /**
+   * Retrieves cached test suites.
+   */
+  def findTestSuites()
+      : CompletableFuture[java.util.List[TestSuiteDiscoveryResult]] =
+    CompletableFuture.supplyAsync(
+      () => {
+        cachedTestSuites.values
+          .groupBy(_.buildTarget)
+          .toSeq
+          .map { case (buildTarget, entries) =>
+            val grouped = groupTestsByPackage(entries.toList)
+            TestSuiteDiscoveryResult(
+              buildTarget.getDisplayName(),
+              buildTarget.getId().getUri(),
+              grouped.asJava
+            )
+          }
+          .asJava
+      },
+      executor
+    )
 
   /**
-   * Find test suites for all build targets in current projects.
+   * Find test suites for all build targets in current projects and update caches.
    */
-  def findTestSuites(): Seq[TestSuiteDiscoveryResult] = {
-    buildTargets.allBuildTargetIds.toList
+  def refreshTestSuites(): Unit = {
+    if (clientConfig.isTestExplorerProvider()) refreshTestSuitesImpl()
+  }
+
+  def refreshTestSuitesImpl(): Unit = executor.execute { () =>
+    val buildTargetList = buildTargets.allBuildTargetIds.toList
       .flatMap(buildTargets.info)
       .filterNot(_.isSbtBuild)
-      .map(buildTarget => createTestDiscoveryFor(buildTarget))
-      // don't send empty results
-      .filter(_.discovered.asScala.nonEmpty)
-  }
 
-  private def createTestDiscoveryFor(buildTarget: BuildTarget) = {
-    val classes = buildTargetClasses
-      .classesOf(buildTarget.getId)
-      .testClasses
-      .toList
-      .flatMap { case (symbol, fullyQualifiedClassName) =>
-        createTestEntry(buildTarget, symbol, fullyQualifiedClassName)
-      }
-    val grouped = groupTestsByPackage(classes)
-
-    TestSuiteDiscoveryResult(
-      buildTarget.getDisplayName,
-      buildTarget.getId.getUri,
-      grouped.asJava
-    )
+    val symbolsPerTarget = buildTargetList
+      .map(buildTarget =>
+        SymbolsPerTarget(
+          buildTarget,
+          buildTargetClasses.classesOf(buildTarget.getId).testClasses
+        )
+      )
+
+    // when test suite is deleted it has to be removed from cache
+    val classNamesFromBsp = symbolsPerTarget.flatMap(_.testSymbols.values).toSet
+    val classNamesToRemove = cachedTestSuites.keySet -- classNamesFromBsp
+    classNamesToRemove.foreach { removedClass =>
+      cachedTestSuites.remove(removedClass)
+    }
+
+    val time0 = System.currentTimeMillis()
+    for {
+      SymbolsPerTarget(buildTarget, testSymbols) <- symbolsPerTarget
+      (symbol, fullyQualifiedClassName) <- testSymbols.readOnlySnapshot().toList
+      // IMPORTANT this check is meant to check for class name, not a symbol
+      if !cachedTestSuites.contains(fullyQualifiedClassName)
+    } computeTestEntry(buildTarget, symbol, fullyQualifiedClassName)
+    val time1 = System.currentTimeMillis()
+    scribe.info(s"Cache refresh took: ${time1 - time0}ms")
```

Comment:
Left on purpose

Reply:
That might happen a lot, no need to leave it in since the same can be accomplished by using trace.

Reply:
I meant it's left on purpose for debugging reasons

Reply:
Ok, so the plan is to remove it before the merge?

Reply:
Yup